docs: initial architecture write-up (#5008)
datawire/ambassador: DevDocumentation/ARCHITECTURE.md

## Table of Contents

- [Table of Contents](#table-of-contents)
- [Overview](#overview)
- [Custom Resource Definitions (CRD)](#custom-resource-definitions-crd)
  - [Apiext](#apiext)
  - [Additional Reading](#additional-reading)
- [Emissary-ingress Container](#emissary-ingress-container)
  - [Startup and Busyambassador](#startup-and-busyambassador)
  - [Entrypoint](#entrypoint)
    - [Watch All The Things (Watt)](#watch-all-the-things-watt)
- [Diagd](#diagd)
- [Ambex](#ambex)
- [Envoy](#envoy)
- [Testing Components](#testing-components)
  - [kat-client](#kat-client)
  - [kat-server](#kat-server)

## Overview

Emissary-ingress provides functionality for load balancing, API gateway and ingress controller capabilities using a familiar kubernetes-native approach. This is done by taking advantage of Kubernetes CRDs along with Kubernetes scaling and resiliency to configure Envoy Proxy to handle routing user requests.

Check [this blog post](https://blog.getambassador.io/building-ambassador-an-open-source-api-gateway-on-kubernetes-and-envoy-ed01ed520844) for additional context around the motivations and architecture decisions of Emissary-ingress.

> At the core of Emissary-ingress is Envoy Proxy which provides a lot of configuration and extensions points. Getting this right can be challenging so Emissary-ingress provides Kubernetes Administrators and Developers a cloud-native way to configure Wnvoy using declarative yaml files. Here are the core components of Emissary-Ingress:

Reviewer: ```suggestion
At the core of Emissary-ingress is Envoy Proxy which provides a lot of configuration and extensions points. Getting this right can be challenging so Emissary-ingress provides Kubernetes Administrators and Developers a cloud-native way to configure Envoy using declarative yaml files. Here are the core components of Emissary-Ingress:
```

- CRDs - extend K8s to enable Emissary-ingress's abstractions (*generated yaml*)
- Apiext - A server that implements the Webhook Conversion interface for CRD's (**own container**)  
- Diagd - provides diagnostic ui, translates snapshots/ir into envoy configuration (*in-process*)
- Ambex - gRPC server implementation of envoy xDS for dynamic envoy configration (*in-process*)

> - Envoy Proxy - Proxy that handles routing all user traffic (*in-process)
> - Ambassador Agent - provides connectivity between cluster and Ambassador Cloud.  (*own container*)\

Reviewer: ```suggestion
- Envoy Proxy - Proxy that handles routing all user traffic (*in-process*)
- Ambassador Agent - provides connectivity between cluster and Ambassador Cloud.  (*own container*)
```

## Custom Resource Definitions (CRD)

Kuberenetes allows extending the Kubernetes API through the use of Customer Resource Definitions (aka CRD's) which allow solutions like Emissary-ingress to provide custom abstractions that provide a K8s familiar developer experience through declarative configuration. CRD's provide validation, strong typing, structured data, versioning and are persisted in `etcd` along with the core Kuberenetes resources.

> Emissary-ingress provides a set of CRD's that are applied to a cluster and then are watched by Emissary-ingress. Emissary-ingress then uses the data from these CRD's along with the standard K8's resources (services, endpoints, etc...) to dynamically generate Envoy Proxy configuration. Depending on the version of Emissary-ingress there might be multiple versions of the CRD's that are suppported.

Reviewer 2: ```suggestion
Emissary-ingress provides a set of CRDs that are applied to a cluster and then are watched by Emissary-ingress. Emissary-ingress then uses the data from these CRDs along with the standard K8s resources (services, endpoints, etc...) to dynamically generate Envoy Proxy configuration. Depending on the version of Emissary-ingress there might be multiple versions of the CRDs that are supported.
```

The build system (`make`) uses [controller-gen](https://book.kubebuilder.io/reference/controller-gen.html) to generate the required YAML representation for the customer resources that can be found at `pkg/getambassador.io/crds.yaml`. This file is auto-generated and checked into the respository. This is the file that is applied to a cluster extending the Kubernetes API. If any changes are made to the custom resources then it needs to be re-generated and checked-in as part of your PR. Running `make generate` will trigger the generation of this file and other generated files (`protobufs`) that are checked into the respository as well. If you want to see more about the build process take a look at `build-aux/generate.mk`.

> > **Annotations**:  K8s allows developers to provide Annotations as well to the standard K8s Resources (Services, Ingress, etc...). Annotations were the preferred method of configuring early versions of Emissary-ingress but annotations did not provide validation and can be error prone. However, with the introduction of CRD's these are now the preferred method and annotations are only supported for backwards compatibility. We won't discuss the annotations much here due to this but rather making you aware that they exist.

Reviewer 2: Idk if annotations were preferred but rather the only option available before the advent of CRDs

### Apiext

> Kubernetes provides the ability to have multiple versions of Custom Resources similiar to the core K8s resources but it is only capable of having a single `storage` version that is persisted in `etcd`. Custom Resource Definitions can define a `ConversionWebHook` that Kubernetes will call whenever it receives a version that is not the storage version. It is the job

Reviewer 2: ```suggestion
Kubernetes provides the ability to have multiple versions of Custom Resources similiar to the core K8s resources but it is only capable of having a single `storage` version that is persisted in `etcd`. Custom Resource Definitions can define a `ConversionWebHook` that Kubernetes will call whenever it receives a version that is not the storage version.
```

> Kubernetes provides the ability to have multiple versions of Custom Resources similiar to the core K8s resources but it is only capable of having a single `storage` version that is persisted in `etcd`. Custom Resource Definitions can define a `ConversionWebHook` that Kubernetes will call whenever it receives a version that is not the storage version. It is the job

Reviewer: ```suggestion
Kubernetes provides the ability to have multiple versions of Custom Resources similiar to the core K8s resources but it is only capable of having a single `storage` version that is persisted in `etcd`. Custom Resource Definitions can define a `ConversionWebHook` that Kubernetes will call whenever it receives a version that is not the storage version. 
```

You can check the current storage version by looking at `pkg/getambassador.io/crds.yaml` and searching for the `storage: true` field and seeing which version is the storage version of the custom resource (*at the time of writing this it is `v2`*).

> The `apiext` container is the Emissary-ingress's server implementation for the conversion webhook that is registered with our custom resources. Each custom resource will have a section similiar to the following in the `pkg/getambassador.io/crds.yaml`:

Reviewer 2: ```suggestion
The `apiext` container is Emissary-ingress's server implementation for the conversion webhook that is registered with our custom resources. Each custom resource will have a section similar to the following in `pkg/getambassador.io/crds.yaml`:
```

```yaml
conversion:
    strategy: Webhook
    webhook:
      clientConfig:
        service:
          name: emissary-apiext
          namespace: emissary-system
      conversionReviewVersions:
      - v1beta1
```

> This is telling the Kubernetes API Server to call a WebHook using a `Service` within the cluster that is called `emissaryapiext` that can be found in `emissary-system` namespace. It also states that our server implementation as of writing this guide only supports `v1beta1` of the WebHook so the K8s API Server will send the request and expect the response in the format for `v1beta1`.

Reviewer 2: ```suggestion
This is telling the Kubernetes API Server to call a WebHook using a `Service` within the cluster that is called `emissary-apiext` that can be found in the `emissary-system` namespace. It also states that our server implementation as of writing this guide only supports `v1beta1` of the WebHook so the K8s API Server will send the request and expect the response in the format for `v1beta1`.
```

The implementation of the `apiext` server can be found in `cmd/apiext` and it leverages the [controller-runtime](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.2starts) library which is vendored in `vendor/sigs.k8s.io/controller-runtime`. When this process starts up it will do the following:

1. Register the Emissary-Ingress CRD schemas using the Go structs described previously
2. Ensure a self-signed certificate is generated that our server can register with for `https`.
3. Kick off a Go routine that handles watching our CRD's and enriching the WebHook Conversion section (*outlined in yaml above*) so that it includes our self-signed certs, port and path that the apiext server is listening on.

> 4. Starts up our two servers one for container liveness/probes and one for the WebHook implementation that performs the conversion between CRD versions.

Reviewer 2: ```suggestion
4. Starts up our two servers one for container liveness/readiness probes and one for the WebHook implementation that performs the conversion between CRD versions.
```

This section will give a high-level overview of each of these components and will help provide you direction on where you can find more information on each of the components.

### Startup and Busyambassador

Emissary-ingress has evolved over many years, many contributors and many versions of Kubernetes which has led to the internal components being implemented in different programming languages. Some of the components are pre-built binaries like envoy, first-party python programs and first-party golang binaries. To provide a single entrypoint for the container startup the Golang binary called `busyambassador` was introduced.

/buildroot/ambassador/python/entrypoint.sh

The `busyambassador` binary provides a busybox like interface that dispatches the CMD's that are provided to a container for the various configured Golang binaries. This enables a single image to support multiple binaries on startup that are declartively set within a `deployment` in the `command` field when setting the image for a deployment. An example of this can be seen in the `ambassador-agent` deployment.

> The image takes advantage of the `ENTRYPOINT` and `CMD` fields within a docker image manifest. You can see this in `builder/Dockerfile` in the final optimized image on the last line there is `ENTRYPOINT [ "bash", "/buildroot/ambassador/python/entrypoint.sh" ]`. This entrypoint cannot be overriden by the user and will run that bash script. By default the bash script will run the `entrypoint` binary which will be discussed in the next section but if passed a known binary name such as the `agent` example then `busyambassador` will run the correct command.

Reviewer 2: you can change entrypoint in a kubernetes config so saying that it can't be changed is not entirely accurate.

| `demomode` (*if enabled*)                                                 | :white_check_mark: |                    |
| `diagd` - admin ui & config processor                                     |                    | :white_check_mark: |
| `ambex` - the Envoy ADS Server                                            | :white_check_mark: |                    |
| `envoy` - proxy routing data                                              |                    | :white_check_mark: |
| SnapshotServer  - expose in-memory snapshot over localhost                | :white_check_mark: |                    |
| ExternalSnapshotServer - Ambassador Cloud friendly exposed over localhost | :white_check_mark: |                    |
| HealthCheck -  endpoints for K8s liveness/readiness probes                | :white_check_mark: |                    |
| Watt - Watch k8s, consul & files for cluster changes                      | :white_check_mark: |                    |
| Sidecar Processes - start various side car processes                      |                    | :white_check_mark: |

> Some of these items will be discussed in more detail but the best places to get started looking at the `entrypoint` is by looking at the `cmd/entrypoint/entrypoint.go`.

Reviewer 2: ```suggestion
Some of these items will be discussed in more detail but the best places to get started looking at the `entrypoint` is by looking at `cmd/entrypoint/entrypoint.go`.
```

The watching of the K8s Cluster changes is where Emissary-ingress will get most of its configuration by looking for K8s Resources (e.g. services,ingresss, etc...) as well as the Emissary-ingress CRD Resources (e.g. Host, Mapping, Listeners, etc...). A `consulWatcher`  will be started if a user has configured a Mapping to use the `ConsulResolver`. You can find this code in `cmd/entrypoint/consul.go`. The filesystem is also watched for changes to support `istio` and how it mounts certificates to the filesystem.

Here is the general flow:

```mermaid
flowchart TD
  subgraph watchers
    K8s & consulW(Consul) & file(FileSystem)
    K8s -.ConsulResolver.-> consulW
    K8s & consulW(Consul) & file(FileSystem) --> sgen
  end
  API[K8s API] --> K8s
  consul[Consul Manager] --> consulW
  istio[Pod Filesystem] --> file

  sgen("Snapshot \n Generation") --> update("Update Snapshot \n(in-memory)") --> diagd(Notify diagD);
  update -.isEdgeStack?.-> AES(Notify AES)
```

## Diagd

Provides two main functions:

> 1. a Diagnostic Admin UI for viewing the current state of Emissary-Ingress

Reviewer: ```suggestion
1. A Diagnostic Admin UI for viewing the current state of Emissary-Ingress
```

2. Processing Cluster changes into Envoy ready configuration
   1. This process has all the steps i'm outlining below

  - receives "CONFIG" event and pushes on queue
  - event queue loop listens for commands and pops them off
  - on CONFIG event it calls back to emissary Snapshot Server to grab current snapshot stored in-memory
  - It is serialized and stored in `/ambassador/snapshots/snapshot-tmp.yaml`.
  - A SecretHandler and Config is initialized
  - A ResourceFetcher (aka, parse the snapshot into an in-memory representation)
  - Generate IR and envoy configs (load_ir function)
    - Take each Resource generated in ResourceFetcher and add it to the Config object as strongly typed objects
    - Store Config Object in `/ambassador/snapshots/aconf-tmp.json`
    - Check Deltas for Mappings cach and determine if we needs to be reset

>     - Create IR with aConfig, Cache, and invalidated items

Reviewer: ```suggestion
    - Create IR with a Config, Cache, and invalidated items
```

      - Rotating them allows for seeing the history of snapshots up to a limit and then they are dropped
      - this also renames the `-tmp` files written above into 
    - Persist bootstrap, envoy ads config and clustermap config to base directory:
      - `/ambassador/bootstrap-ads.json` # this is used by envoy during startup to initial config itself and let it know about the static ADS Service
      - `/ambassador/enovy/envoy.json` # this is used in `ambex` to generate the ADS snapshots along with the fastPath items
      - `/ambassador/clustermap.json` # this might not be used either...
    - Notify `envoy` and `ambex` that a new snapshot has been persisted using signal SIGHUP
      - the Goroutine within `entrypoint` that starts up `envoy` is blocking waiting for this signal to start envoy
      - the `ambex` process continuously listens for this signal and it triggers a configuration update for ambex.
    - Update the appropriate status fields with metatdata by making calls to the `kubestatus` binary found in `cmd/kubestatus` which handles the communication to the cluster

## Ambex

This is the gRPC server implementation of the envoy xDS v2 and v3 api's based on ...

> - listens for SIGUP from diagd

Reviewer: ```suggestion
- listens for SIGHUP from diagd
```
Not sure about this one, but just referencing the above

- converts `envoy.json` into in-memory snapshots that are cached for v2/v3
- implements ADS v2/v3 Apis that envoy is configured to listen to

## Envoy

> We have our own forked version that is behind and has a few one off commits.

Reviewer 2: ```suggestion
We maintain our own [fork](https://github.com/datawire/envoy) of Envoy that includes some additional commits for implementing some features in Emissary-Ingress.
```
